logstash-core/lib/logstash/java_pipeline.rb

```diff
@@ -309,9 +316,17 @@ def resolve_cluster_uuids
     end.to_a.compact
   end
 
-  def wait_inputs
-    @input_threads.each do |thread|
-      thread.join # Thread or java.lang.Thread (both have #join)
+  def monitor_inputs_and_workers
+    twait = ThreadsWait.new(*(@input_threads + @worker_threads))
+
+    while !@input_threads.empty?
+      terminated_thread = twait.next_wait
+      if @input_threads.delete(terminated_thread).nil?
+        # this is a worker thread termination
+        # delete it from @worker_threads so that wait_for_workers does not wait for it
+        @worker_threads.delete(terminated_thread)
+        raise("Worker thread terminated in pipeline.id: #{pipeline_id}")
```

Comment:
So 2 things:
- Having a worker thread exception has always been considered a fatal at pipeline level. In the single pipeline scenario a worker thread exception has always been stopping logstash altogether. Before this PR a single worker exception in a multi pipeline scenario would ALSO crash logstash completely but that does not make sense and instead only the pipeline with the failed worker should be terminated. 
- When this is raised here, it is rescue'd in the `start` method. The calling chain is `start` -> `run` -> `monitor_inputs_and_workers`.  
https://github.com/elastic/logstash/blob/e99f37dc7b863ea23686ebfc4b8f926e969aaa23/logstash-core/lib/logstash/java_pipeline.rb#L123-L137

Reply:
I believe that raising here is correct in the sense that -it is- an exception to have a worker termination and that exception will be propagated and dealt with up the stack accordingly, otherwise a worker termination will simply terminate the pipeline in a non-exceptional way, just like when the inputs terminate which is not IMO correct.